best-practices/largescale-deep-learning/Training/ViT-Pretrain/README.md

```diff
@@ -0,0 +1,106 @@
+5. Scale up before scaling out, e.g. use a Standard_NC24s_v3 with 4 GPUs instead of 4 Standard_NC6s_v3's with 1 GPU each [NCv3 machines](https://learn.microsoft.com/en-us/azure/virtual-machines/ncv3-series). Synchronizing weights on one machine with 4 GPUs (Standard_NC24s_v3) is faster than synchronizing weights across the network between 4 machines with one GPU each (Standard_NC6s_v3).
+6. Use [InfiniBand]#(hardware) whenever possible, as it enables near linear scaling.
+7. If working with high resolution images, strongly consider resizing them or extracting tiles from them before training. If training a ViTB model, images larger than 2048 x 2048 will lead to severe bottlenecks in CPU processing, reading from disk and transferring data over the network.
+8. Depending on your particular application, the [Nvidia DALI](https://developer.nvidia.com/dali) package might help reduce the load on the CPU by doing image preprocessing and augmentation on the GPU.
+
+## **Experiment Setup**
+
+We derived some of the ideas above via experimentation. Since our experiment setup can be a good starting point for implementing large scale training for Computer Vision, we describe it below.
+
+### **Hardware**
+
+We used machines with [V100](https://learn.microsoft.com/en-us/azure/virtual-machines/ncv3-series) and [A100](https://learn.microsoft.com/en-us/azure/virtual-machines/nc-a100-v4-series) GPUs. We also used InfiniBand, which is specialized hardware for linear scaling. InfiniBand enables low-latency, GPU-to-GPU communication across nodes in a cluster. You can view the full list of InfiniBand-enabled Azure machine SKUs [here](https://learn.microsoft.com/en-us/azure/virtual-machines/sizes-hpc#rdma-capable-instances).
+
+### **Code**
+
+All of the code in our experiments uses the PyTorch Lightning framework and can be found in the ``src`` subdirectory of this directory.
+
+### **Job Submission**
+
+The code assumes that the workspace where the job runs contains a datastore named `datasets` with a blob named `imagenet`. The folders under it must be readable by PyTorch's `ImageFolder` class. For more details, please see the [launcher.py](launcher.py) and [run_image_classification.py](src/run_image_classification.py) files.
+
+To train a ViTB model on ImageNet-1k, submit the following command from within the ``src`` subdirectory of this directory:
+```
+python launcher.py
+```
+
+## **Experiment Results**
+
+We show the results of our experiments below.
+
+### **Multiple nodes / multiple GPUs**
+
+We trained ViTB on ImageNet-1k in a few hardware configurations with varying numbers of nodes and GPUs. The GPU used was V100.
+
+| Nodes | GPUs per Node | Total GPUs | InfiniBand | Training Speed (images/second) |
+|-------|---------------|------------|------------|--------------------------------|
+| 1 | 1 | 1 | no | 101.6 |
+| 1 | 4 | 4 | no | 376.3 |
+| 1 | 4 | 4 | yes | 377.3 |
```

Comment:
Should we remove row 3. Its not adding any value here.

Reply:
Do you mean we already knew that Infiniband on one machine does not make a difference? If we exclude this data point, the Infiniband line in the graph below would start at 16, which would be odd, so I'd rather keep things as they are.

Reply:
keeping the row could also confuse people that infiniband is not helping improve the performance.

Reply:
Makes sense. Can clarify this in a note.